### PR TITLE
cephadm: list-networks: Avoid duplicated IPs

### PR DESCRIPTION
**Pull request by sebastian-philipp**
Fixes: https://tracker.ceph.com/issues/52083
Signed-off-by: Sebastian Wagner <sewagner@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by liewegas — APPROVED** [review]

**Comment by fmount** [comment]
Overall the patch looks good to me and using Set instead of List can easily solve avoiding duplicates.
We can merge this change or, if a -pending container (including this change) is built, we can try it triggering the OSP CI.
+1 on my side

**Comment by fmount** [comment]
After a few more testing (on a standalone Ceph cluster, but it's pretty much the same of the OSP CI), I found the following stack trace [1] in the ceph-mgr log, because `set is not JSON serializable`.
In cephadm I see a lot of `json.dumps(...)` within the code, and for a long term solution, I suggest to define a JSONEncoder utility class that can help handling different (or mixed) data structures.
In order to make this PR working properly I propose to add a few more code like [2], but feel free to make your tests and come up with another solution (e.g. you can avoid using a JSONEncoder that explicitly defines the logic of translating data structures and reach the json.dumps() with the proper, serializable data). 

[1] https://paste.opendev.org/show/808195/
[2] https://paste.opendev.org/show/808196/

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -4823,7 +4823,7 @@ def command_logs(ctx):
 
 
 def list_networks(ctx):
-    # type: (CephadmContext) -> Dict[str,Dict[str,List[str]]]
+    # type: (CephadmContext) -> Dict[str,Dict[str, Set[str]]]
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:4826]:
```
  File "/var/lib/ceph/4b5c8c0a-ff60-454b-a1b4-9747aa737d19/cephadm.037f1495e23201a98a9b38d802e2e3f0623e9cc83d25f40b0ce9350dcbc53442", line 8230, in <module>
    main()
  File "/var/lib/ceph/4b5c8c0a-ff60-454b-a1b4-9747aa737d19/cephadm.037f1495e23201a98a9b38d802e2e3f0623e9cc83d25f40b0ce9350dcbc53442", line 8218, in main
    r = ctx.func(ctx)
  File "/var/lib/ceph/4b5c8c0a-ff60-454b-a1b4-9747aa737d19/cephadm.037f1495e23201a98a9b38d802e2e3f0623e9cc83d25f40b0ce9350dcbc53442", line 4737, in command_list_networks
    print(json.dumps(r, indent=4))
  File "/usr/lib64/python3.6/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
  File "/usr/lib64/python3.6/json/encoder.py", line 201, in encode
    chunks = list(chunks)
  File "/usr/lib64/python3.6/json/encoder.py", line 430, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/usr/lib64/python3.6/json/encoder.py", line 404, in _iterencode_dict
    yield from chunks
  File "/usr/lib64/python3.6/json/encoder.py", line 404, in _iterencode_dict
    yield from chunks
  File "/usr/lib64/python3.6/json/encoder.py", line 437, in _iterencode
    o = _default(o)
  File "/usr/lib64/python3.6/json/encoder.py", line 180, in default
    o.__class__.__name__)
TypeError: Object of type 'set' is not JSON serializable

File: src/cephadm/tests/fixtures.py [hunk 2]
```diff
@@ -106,8 +104,12 @@ def with_cephadm_ctx(
...
          mock.patch('cephadm.json_loads_retry', return_value={'epoch' : 1}), \
-         mock.patch('cephadm.list_networks', return_value=list_networks), \
          mock.patch('socket.gethostname', return_value=hostname):
-             ctx: cd.CephadmContext = cd.cephadm_init_ctx(cmd)
-             ctx.container_engine = container_engine
-             yield ctx
+        ctx: cd.CephadmContext = cd.cephadm_init_ctx(cmd)
+        ctx.container_engine = container_engine
+        if list_networks is not None:
+            with mock.patch('cephadm.list_networks', return_value=list_networks):
```
Comment (mgfritch) [inline src/cephadm/tests/fixtures.py:111]: hmm would it be better to add a `mock.patch` decorator to each unittest where it's used instead?

**Review by mgfritch — APPROVED** [review]

File: src/cephadm/cephadm [hunk 3]
```diff
@@ -4906,15 +4905,19 @@ def _parse_ipv6_route(routes: str, ips: str) -> Dict[str, Dict[str, List[str]]]:
...
 
     return r
 
 
 def command_list_networks(ctx):
     # type: (CephadmContext) -> None
     r = list_networks(ctx)
-    print(json.dumps(r, indent=4))
+
+    def serialize_sets(obj: Any) -> Any:
```
Comment (fmount) [inline src/cephadm/cephadm:4917]: even better, LGTM, thanks @sebastian-philipp !!

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2021-08-24_13:38:10-rados:cephadm-wip-swagner-testing-2021-08-24-1031-distro-basic-smithi/

**Comment by fmount** [comment]
@sebastian-philipp are you going to backport this change in the pacific branch ?

**Pull request merged**